packages/core/src/fhir-to-cda/cda-templates/components/observations.ts

```diff
@@ -1,37 +1,251 @@
+    const componentTrs = observation.resource.component.flatMap(pair => {
+      pairNumber++;
+      return createTableRowFromObservation(
+        pair,
+        buildReferenceId(socHistPrefix, pairNumber),
+        formattedDate
+      );
+    });
+    trs.push(...componentTrs);
+  }
+  if (hasObservationInCode(observation.resource)) {
+    pairNumber++;
+    trs.push(
+      createTableRowFromObservation(
+        observation.resource,
+        buildReferenceId(socHistPrefix, pairNumber),
+        formattedDate
+      )
+    );
+  }
+  return trs;
+}
+
+export function createTableRowFromObservation(
+  observation: Observation | ObservationComponent,
+  referenceId: string,
+  date: string | undefined
+): ObservationTableRow {
+  const interpretation = observation.interpretation;
+  const score = interpretation?.[0]?.coding?.[0]?.display;
+  const intValue = score ? parseInt(score) : undefined;
+  const scoreValue = intValue != undefined && !isNaN(intValue) ? intValue.toString() : "N/A";
+
+  return {
+    tr: {
+      [_idAttribute]: referenceId,
+      ["td"]: [
+        {
+          [_inlineTextAttribute]: observation.code?.coding?.[0]?.display ?? observation.code?.text,
+        },
+        {
+          [_inlineTextAttribute]:
+            observation.valueCodeableConcept?.coding?.[0]?.display ??
+            observation.valueCodeableConcept?.text ??
+            "Not on file",
+        },
+        {
+          [_inlineTextAttribute]: scoreValue,
+        },
+        {
+          [_inlineTextAttribute]: date ?? "Unknown",
+        },
+      ],
+    },
+  };
+}
+
+function hasObservationInCode(observation: Observation): boolean {
+  return (
+    (observation.code?.coding &&
+      observation.code.coding.length > 0 &&
+      observation.valueCodeableConcept?.coding &&
+      observation.valueCodeableConcept.coding.length > 0) ??
+    false
+  );
+}
+
+export function createEntriesFromObservation(
+  observation: AugmentedObservation,
+  socHistNumber: string
+): ObservationEntry[] {
+  let pairNumber = 0;
+  const date = formatDateToCdaTimestamp(observation.resource.effectiveDateTime);
+  const observationEntry = createEntryFromObservation(
+    observation.resource,
+    observation,
+    socHistNumber,
+    date
+  );
+  observationEntry.observation.entryRelationship = [];
+
+  if (observation.resource.component && observation.resource.component.length > 0) {
+    observation.resource.component.map(pair => {
+      pairNumber++;
+      const entryRelationshipObservation = createEntryFromObservation(
+        pair,
+        observation,
+        buildReferenceId(socHistNumber, pairNumber),
+        date
+      );
+      const entryRelationship = createEntryRelationship(entryRelationshipObservation);
+      if (observationEntry.observation.entryRelationship)
+        observationEntry.observation.entryRelationship.push(entryRelationship);
+    });
+  }
+
+  return [observationEntry];
```

Comment:
ditto re returning an array of 1 object

Reply:
This allows the generic `createTableRowsAndEntries` function to reuse different callbacks - some of which return true arrays, while others return arrays of one item. Could improve this with that array-checking function we recently talked about on one of your PRs